Implements isOpen() / close() methods on RemoteCollection

reinhapa opened the pull request:
### Description:
Implements the `Collection.isOpen()` method behaviour and also moved duplicated `XmlRpcClient` acces into a central method on the `RemoteCollection` eliminating repetitive implementations all over multiple places reducing the total amount of code.

### Type of tests:
Adds a new JUnit tests for the new implemented parts as the enhanced parts from `Leasable`

dizzzz commented:
@duncdrum there is a docker issue (java11): any idea?

<img width="950" alt="image" src="https://user-images.githubusercontent.com/1700062/61323530-8670e700-a810-11e9-9a12-8bb205046968.png">

dizzzz commented on exist-core/src/main/java/org/exist/xmldb/AbstractRemoteResource.java:
```diff
@@ -156,11 +154,7 @@ public void setLastModificationTime(final Date dateModified) throws XMLDBExcepti
-            } catch (final XmlRpcException e) {
-                throw new XMLDBException(ErrorCodes.UNKNOWN_ERROR, e.getMessage(), e);
-            }
+            collection.execute("setLastModified", params);
```
should we make this string a (shared) constant?

dizzzz replied:
nope, there are too many of them?

reinhapa replied:
@dizzzz you mean `"setLastModified"`? Where should that be put to?

dizzzz replied:
yes. but forget about my remark as there are soo many, and... they are only used once right?

reinhapa replied:
yes

dizzzz commented on exist-core/src/main/java/org/exist/xmldb/AbstractRemoteResource.java:
```diff
@@ -487,20 +479,16 @@ protected long getStreamLengthInternal(final Object obj)
-                    } else {
-                        retval = Long.parseLong((String) o);
-                    }
+            final Map<?, ?> table = (Map<?, ?>) collection.execute("describeResource", params);
```
shared constant?

dizzzz commented:
@adamretter do you want to have this in 5.x as well?

dizzzz commented:
please could you rebase ? the docker issues are biting the PR :-(

adamretter commented on exist-core/src/main/java/org/exist/util/Leasable.java:
```diff
@@ -132,17 +133,30 @@ public boolean isLeased() {
+    public boolean isClosed() {
+        lock.readLock().lock();
+        try {
+            return Leasable.this.closed;
```
I think just `return closed;` is fine.

adamretter commented on exist-core/src/main/java/org/exist/xmldb/RemoteCollection.java:
```diff
@@ -110,37 +110,47 @@ public static RemoteCollection instance(final Leasable<XmlRpcClient> leasableXml
+
     @Override
     public void close() {
+        if (xmlRpcClientLease.isClosed()) {
```
I don't think this is required as calling closed is `idempotent`

reinhapa replied:
No, because subsequent call on xmlRpcClientLease.close() will throw an error, that I don't want to have here. The `close()` method should gracefully ignore subsequent calls here.

adamretter commented:
I think it would be useful to have a call to discuss the design of this ;-)

reinhapa commented:
@adamretter we could have a call to discuss this change any time you like...

adamretter commented:
@reinhapa sorry for the delay. The scale of this change has put me off a little. I will try and find some time for it ASAP

reinhapa commented:
@adamretter I will check if I can find all the tests that will need to close the collections. Also I found only the following classes where I added the `RemoteCollection` for the simplification of the remote call so far (maybe I missed an other one):
`RemoteDatabaseInstanceManager`
`RemoteRestoreService`
for those I introduced a `RemoteCallSite` functional interface that specifies the needed method signature and does so hides away the actual needed RPC client implementation. There is only one spot for changing the password, where I would suggest to introduce a "virtual" method name for changing the password in order to remove the need for passing in the client reference, but I think I will do that in a next PR.

duncdrum commented:
i have the feeling that we'll only know the true extend on missing `close` calls, if we move this one ahead, to give the wider community a chance to locate problem areas.

reinhapa commented:
I will rebase the source in order to tidy up all changes into one commit

dizzzz commented on exist-core/src/main/java/org/exist/xmldb/RemoteXMLResource.java:
```diff
@@ -437,11 +429,7 @@ public void setDocType(final DocumentType doctype) throws XMLDBException {
-            } catch (final XmlRpcException e) {
-                throw new XMLDBException(ErrorCodes.UNKNOWN_ERROR, e.getMessage(), e);
-            }
+            collection.execute("setDocType", params);
```
maybe we can share `"setDocType"` as magic string in collection?

dizzzz commented:
@adamretter please could you mark your concerns on 'resolved' if you agree with the answers?

adamretter commented:
@reinhapa thanks for your contribution and sorry for the delays.
